src/Admin/SharedBlockAdmin.php

```diff
@@ -80,15 +83,44 @@ protected function configureFormFields(FormMapper $formMapper)
 
         $formMapper->with('form.field_group_options');
 
-        /** @var BaseBlockService $service */
+        $this->configureBlockFields($formMapper, $block);
+
+        $formMapper->end();
+    }
+
+    private function configureBlockFields(FormMapper $formMapper, BlockInterface $block): void
+    {
         $service = $this->blockManager->get($block);
 
-        if ($block->getId() > 0) {
-            $service->buildEditForm($formMapper, $block);
-        } else {
-            $service->buildCreateForm($formMapper, $block);
+        if (!$service instanceof BlockServiceInterface) {
+            throw new \RuntimeException(sprintf(
+                'The block "%s" is not a valid %s',
+                $block->getType(),
+                BlockServiceInterface::class
+            ));
         }
 
-        $formMapper->end();
+        if ($service instanceof EditableBlockService) {
+            $blockMapper = new PageFormMapper($formMapper);
+            if ($block->getId() > 0) {
+                $service->configureEditForm($blockMapper, $block);
+            } else {
+                $service->configureCreateForm($blockMapper, $block);
+            }
+        } else {
+            @trigger_error(
+                sprintf(
+                    'Editing a non %s block service is deprecated since sonata-project/page-bundle 3.x and will not be allowed with version 4.0.',
```

Comment:
```suggestion
                    'Editing a block service that doesn\'t implement %s is deprecated since sonata-project/page-bundle 3.x and will not be allowed with version 4.0.',
```